crates/holochain_types/src/test_utils/chain.rs

```diff
@@ -107,31 +193,80 @@ pub fn gap_chain(ranges: &[Range<u32>]) -> Vec<TestChainItem> {
         .collect()
 }
 
+/// Produce an arbitrary SignedActionHashed from any ChainItem.
+///
+/// The SignedActionHashed will not be valid in any sense other than the
+/// fields relevant to ChainItem.
+pub fn chain_item_to_action(u: &mut Unstructured, i: &impl ChainItem) -> SignedActionHashed {
+    let action_seq = i.seq();
+    let prev_action = i.prev_hash().cloned().map(Into::into);
+    let hash: ActionHash = i.get_hash().clone().into();
+    let mut action = SignedActionHashed::arbitrary(u).unwrap();
+    match (action_seq, prev_action) {
+        (_, None) => {
+            let dna = Dna::arbitrary(u).unwrap();
+            action.hashed.content = Action::Dna(dna);
+            action.hashed.hash = hash;
+        }
+        (action_seq, Some(prev_action)) => {
+            let mut create = Create::arbitrary(u).unwrap();
+            create.action_seq = action_seq;
+            create.prev_action = prev_action;
+            action.hashed.content = Action::Create(create);
+            action.hashed.hash = hash;
+        }
+    }
+    action
+}
+
+/// Produce a sequence of AgentActivity ops from a Vec of ChainItems
 pub fn chain_to_ops(chain: Vec<impl ChainItem>) -> Vec<RegisterAgentActivity> {
     let mut u = Unstructured::new(&holochain_zome_types::NOISE);
     chain
         .into_iter()
         .map(|i| {
-            let action_seq = i.seq();
-            let prev_action = i.prev_hash().cloned().map(Into::into);
-            let hash: ActionHash = i.get_hash().clone().into();
             let mut op = RegisterAgentActivity::arbitrary(&mut u).unwrap();
-            match (action_seq, prev_action) {
-                (0, _) => {
-                    let dna = Dna::arbitrary(&mut u).unwrap();
-                    op.action.hashed.content = Action::Dna(dna);
-                    op.action.hashed.hash = hash;
-                }
-                (action_seq, Some(prev_action)) => {
-                    let mut create = Create::arbitrary(&mut u).unwrap();
-                    create.action_seq = action_seq;
-                    create.prev_action = prev_action;
-                    op.action.hashed.content = Action::Create(create);
-                    op.action.hashed.hash = hash;
-                }
-                _ => unreachable!(),
-            }
+            op.action = chain_item_to_action(&mut u, &i);
             op
         })
         .collect()
 }
+
+isotest::iso! {
+    TestChainItem => |i| {
+        let mut u = Unstructured::new(&holochain_zome_types::NOISE);
+        chain_item_to_action(&mut u, &i)
+    },
+    SignedActionHashed => |a| {
+        TestChainItem {
+            seq: a.seq(),
+            hash: TestChainHash::test(a.get_hash()),
+            prev: a.prev_hash().map(TestChainHash::test),
+        }
+    },
+    test_cases: [
```

Comment:
The `iso!` macro allows you to specify test cases for which it will automatically generate unit tests which check that certain idempotency invariants are upheld, i.e. that the Iso trait is implemented correctly in that it doesn't lose information during conversions.